src/main.rs

```diff
@@ -223,3 +519,108 @@ async fn main() -> zbus::Result<()> {
         })
         .await
 }
+
+async fn watch_config_message_stream(
+    conn: Connection,
+    watched_configs: Arc<
+        RwLock<HashMap<(String, u64), (Connection, ObjectPath<'static>, WellKnownName<'static>)>>,
+    >,
+    watched_states: Arc<
+        RwLock<HashMap<(String, u64), (Connection, ObjectPath<'static>, WellKnownName<'static>)>>,
+    >,
+) -> zbus::Result<()> {
+    let config_rule = MatchRule::builder()
+        .msg_type(zbus::MessageType::MethodCall)
+        .member("WatchConfig")?
+        .interface("com.system76.CosmicSettingsDaemon")?
+        .build();
+    let config_stream = MessageStream::for_match_rule(config_rule, &conn, Some(100)).await?;
+
+    let mut watched_config_names: HashMap<(String, u64), HashSet<UniqueName<'static>>> =
+        HashMap::new();
+
+    let state_rule = MatchRule::builder()
+        .msg_type(zbus::MessageType::MethodCall)
+        .member("WatchState")?
+        .interface("com.system76.CosmicSettingsDaemon")?
+        .build();
+    let state_stream = MessageStream::for_match_rule(state_rule, &conn, Some(100)).await?;
+
+    let mut watched_state_names: HashMap<(String, u64), HashSet<UniqueName<'static>>> =
+        HashMap::new();
+
+    let name_changed_rule = MatchRule::builder()
+        .msg_type(zbus::MessageType::Signal)
+        .sender("org.freedesktop.DBus")?
+        .member("NameOwnerChanged")?
+        .interface("org.freedesktop.DBus")?
+        .arg(2, "")? // new owner is empty
+        .build();
+
+    let name_changed_stream =
+        MessageStream::for_match_rule(name_changed_rule, &conn, Some(100)).await?;
+
+    let mut rx = name_changed_stream.merge(config_stream).merge(state_stream);
+
+    while let Some(msg) = rx.try_next().await? {
+        if msg.member() == Some(MemberName::from_static_str_unchecked("NameOwnerChanged")) {
```

Comment:
Is there a reason this isn't just using https://docs.rs/zbus/3.14.1/zbus/fdo/struct.DBusProxy.html#method.receive_owner_changed?

Reply:
Ya, there were a couple reasons. One was that I could add a filter for the argument so that only messages where the new owner is empty are received and another was just because it was easier to merge with the other streams :sweat_smile: